### PR TITLE
[Fix #3937] Add new Rails/StartEndWith cop

### LINKED ISSUE #3937: Implement a cop enforcing a consistent use of ActiveSupport aliases
`ActiveSupport` defines a bunch of core lib method aliases (such as `String#starts_with?` and so on). 
In most Rails projects people end up with a mixed usage of the built-in method names and the ActiveSupport aliases. It'd be nice if we had a cop to enforce using either the AS aliases or the core lib method names consistently.

This cop would be pretty easy to implement and would be a great opportunity for someone to dive into the RuboCop codebase.

### PR DESCRIPTION
**Pull request by tdeo**
This cop checks for consistent use of `starts_with?` and `ends_with?` string
methods or consistent use of `start_with?` and `end_with?`

No option to mix consistent use of `starts_with?` and `end_with?` or the opposite.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [ ] All tests are passing.
* [ ] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

### CODE REVIEW

**Comment by bbatsov** [comment]
We should actually have a single cop checking for all similar `ActiveSupport` aliases.

**Comment by tdeo** [comment]
Do you have any specific one in mind? Or how could we enumerate all such singular / plural forms in aliases?

**Comment by bbatsov** [comment]
Here's a listing of all the methods - http://guides.rubyonrails.org/active_support_core_extensions.html

**Comment by tdeo** [comment]
There's indeed several other examples: `String#camelize` and `String#titleize`aliased respectively to `String#camelcase` and `String#titlecase`.

Otherwise there's also:
- `Array#append` and `Array#prepend`
- `Hash#reverse_update`, `Hash#to_options` and `Hash#to_options!`
- Several `Date` and `DateTime` methods.
- The `Numeric#byte`, `Numeric#kilobyte`, `Numeric#megabyte`, ... available in singular and plural forms (which I fell should belong in Rails/PluralizationGrammar)

I'm not sure if all of them should belong to the same cop (which would then be `Rails/ActiveSupportAliases` that can be enabled / disabled) ?

**Comment by rrosenblum** [comment]
Aren't some of these aliases designed for grammatical purposes more so than as a replacement? In my mind this should work similar to what is being done in `Rails/PluralizationGrammar`. Rather than enforcing an always use one style, I would expect this to attempt to figure out if the receiver uses a singular or a plural name.

**Comment by bbatsov** [comment]
> Aren't some of these aliases designed for grammatical purposes more so than as a replacement?

They were, but this is pointless in a programming language. Matz himself mentioned this a while back when asked what he thinks about those.

**Comment by tdeo** [comment]
So we would rather head towards a cop `Rails/VanillaSemantics` that would handle all of them at once?

People will still have a choice to enable or disable it anyway. But yeah, it would be "all or nothing".

**Comment by tdeo** [comment]
@bbatsov I renamed the cop to `Rails/ActiveSupportAliases` to be consistent with https://github.com/bbatsov/rubocop/pull/3935. 

I'm only checking for `String#starts_with?`, `String#ends_with?`, `Array#append` and `Array#prepend` which are the only aliases to core ruby methods (the other I mentioned before are actually alias to other ActiveSupport methods). 

A detail here: I'm not supporting autocorrect from `Array#append` to `Array#<<` because I didn't know how to handle removing the `.` for method call and the parameter parenthesis if present. I felt that autocorrecting to `array.<<(element)` was too ugly.

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +16,7 @@
 * [#4018](https://github.com/bbatsov/rubocop/pull/4018): Add autocorrect `Lint/EmptyEnsure` cop. ([@pocke][])
 * [#4028](https://github.com/bbatsov/rubocop/pull/4028): Add new `Style/IndentHeredoc` cop. ([@pocke][])
 * Add new `Style/InverseMethods` cop. ([@rrosenblum][])
+# [#3937](https://github.com/bbatsov/rubocop/issues/3937): Add new `Rails/ActiveSupportAliases` cop. ([@tdeo][])
```
Comment (bbatsov) [inline CHANGELOG.md]: You have a typo here - # should be *.

File: config/enabled.yml [hunk 2]
```diff
@@ -1503,6 +1503,13 @@ Rails/ActionFilter:
   Description: 'Enforces consistent use of action filter methods.'
   Enabled: true
 
+Rails/ActiveSupportAliases:
+  Description: >-
+                  Avoid ActiveSupport aliases of standard ruby methods:
+                  `String#starts_with?`, `String#ends_with?`,
+                  `Array#append`, `Array#prepend`
```
Comment (bbatsov) [inline config/enabled.yml]: This sentence should end with a `.`.

File: lib/rubocop/cop/rails/active_support_aliases.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Rails
+      # This cop checks that ActiveSupport aliases to core ruby methods
+      # are not used
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/active_support_aliases.rb]: Add `.` here.

File: lib/rubocop/cop/rails/active_support_aliases.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      #
+      # @example
+      #   # good
+      #   'some_string'.start_with?('prefix')
+      #   'some_string'.end_with?('suffix')
+      #   [1, 2, 'a'] << 'b'
+      #   [1, 2, 'a'].unshift('b')
+      #   # bad
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/active_support_aliases.rb]: Add a blank line between the code examples.

**Comment by bbatsov** [comment]
> A detail here: I'm not supporting autocorrect from Array#append to Array#<< because I didn't know how to handle removing the . for method call and the parameter parenthesis if present. I felt that autocorrecting to array.<<(element) was too ugly.

You can simply replace `.` with ` op ` in the autocorrect. It's accessible via `node.loc` as everything else.

I've added a few small remarks to the code. Btw, ideally this cop should be configurable to both enforce consistent non-usage of the aliases and their consistent usage. I'll merge the cop even in its current form, but I guess some people won't be happy about the lack of config options.

**Comment by tdeo** [comment]
I still can't figure a good way for autocorrecting `append` that handles all those cases:
```ruby
array.append(3)
array.append 3
array
  .append 3
```

**Comment by bbatsov** [comment]
Well, let's leave this out of the cop's scope for now.

**Comment by tdeo** [comment]
And about different styles, I don't see any issue adding the option for consistent non-usage, but I think some people will still be unhappy having to write `Array#append` just because you want to use `String#ends_with?`.

**Comment by bbatsov** [comment]
Fine by me. Just squash and rebase and I'll merge this.

**Pull request merged**